pkg/gadgets/execsnoop/tracer/core/tracer.go

```diff
@@ -0,0 +1,208 @@
+
+type Config struct {
+	// TODO: Make it a *ebpf.Map once
+	// https://github.com/cilium/ebpf/issues/515 and
+	// https://github.com/cilium/ebpf/issues/517 are fixed
+	MountnsMap string
+}
+
+type Tracer struct {
+	c             *Config
+	objs          execsnoopObjects
+	enterLink     link.Link
+	exitLink      link.Link
+	resolver      containercollection.ContainerResolver
+	eventCallback func(types.Event)
+	done          chan bool
+	node          string
+}
+
+func NewTracer(c *Config, resolver containercollection.ContainerResolver,
+	eventCallback func(types.Event), node string) (*Tracer, error) {
+	t := &Tracer{c: c}
+
+	t.resolver = resolver
+	t.eventCallback = eventCallback
+	t.done = make(chan bool)
+	t.node = node
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Stop() {
+	t.stop()
+}
+
+func (t *Tracer) stop() {
+	close(t.done)
+
+	if t.enterLink != nil {
+		t.enterLink.Close()
+		t.enterLink = nil
+	}
+	if t.exitLink != nil {
+		t.exitLink.Close()
+		t.exitLink = nil
+	}
+	t.objs.Close()
+}
+
+func (t *Tracer) start() error {
+	spec, err := loadExecsnoop()
+	if err != nil {
+		return fmt.Errorf("Failed to load ebpf program: %w", err)
+	}
+
+	filter_by_mnt_ns := false
+
+	if t.c.MountnsMap != "" {
+		filter_by_mnt_ns = true
+		m := spec.Maps["mount_ns_set"]
+		m.Pinning = ebpf.PinByName
+		m.Name = filepath.Base(t.c.MountnsMap)
+	}
+
+	consts := map[string]interface{}{
+		"filter_by_mnt_ns": filter_by_mnt_ns,
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil {
+		return fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	opts := ebpf.CollectionOptions{
+		Maps: ebpf.MapOptions{
+			PinPath: filepath.Dir(t.c.MountnsMap),
+		},
+	}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		return fmt.Errorf("Failed to load ebpf program: %w", err)
+	}
+
+	enter, err := link.Tracepoint("syscalls", "sys_enter_execve", t.objs.TracepointSyscallsSysEnterExecve)
+	if err != nil {
+		return fmt.Errorf("Error opening tracepoint: %w", err)
+	}
+	t.enterLink = enter
+
+	exit, err := link.Tracepoint("syscalls", "sys_exit_execve", t.objs.TracepointSyscallsSysExitExecve)
+	if err != nil {
+		return fmt.Errorf("Error opening tracepoint: %w", err)
+	}
+	t.exitLink = exit
+
+	go t.run()
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	// TODO: what's the right value for 4096?
+	pb, err := perf.NewReader(t.objs.execsnoopMaps.Events, 4096)
+	if err != nil {
+		t.eventCallback(types.Event{
+			Event: eventtypes.Event{
+				Type:    eventtypes.ERR,
+				Message: fmt.Sprintf("Error creating perf ring buffer: %v", err),
+				Node:    t.node,
+			},
+		})
+		return
+	}
+	defer pb.Close()
+
+	for {
+		record, err := pb.Read()
+		if err != nil {
+			t.eventCallback(types.Event{
+				Event: eventtypes.Event{
+					Type:    eventtypes.ERR,
+					Message: fmt.Sprintf("Error reading perf ring buffer: %v", err),
+					Node:    t.node,
+				},
+			})
+			return
+		}
```

Comment:
IIUC after our discussion, this routine is being stopped when the perf ring buffer is closed. Isn't there any other way to do it so that we have more control over it?

In addition, it would be great if we can add another condition so that we are completely sure this error is not being sent when the gadget is being stopped.

Reply:
It turns out that this was not being closed. The cilium/ebpf library clones the perf ring buffer map (call's dup() on its fd), hence calling `objs.Close()` was not enough to close this map and stop the go routime reading it. I modified the code to explicitly calling `Close()` on the perf ring buffer reader.